program/docs/nikto.1
.\"     Title: nikto
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 01/19/2010
.\"    Manual: 
.\"    Source: 
.\"
.TH "NIKTO" "1" "01/19/2010" "" ""
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
nikto \- Scan web server for known vulnerabilities
.SH "SYNOPSIS"
.HP 21
\fBnikto\fR [options...]
.SH "DESCRIPTION"
.PP
Examine a web server to find potential problems and security vulnerabilities, including:
.sp
.RS 4
\h'-04'\(bu\h'+03'Server and software misconfigurations
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Default files and programs
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Insecure files and programs
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Outdated servers and programs
.RE
.PP
Nikto is built on LibWhisker (by RFP) and can run on any platform which has a Perl environment\&. It supports SSL, proxies, host authentication, IDS evasion and more\&. It can be updated automatically from the command\-line, and supports the optional submission of updated version data back to the maintainers\&.
.SH "OPTIONS"
.PP
Below are all of the Nikto command line options and explanations\&. A brief version of this text is available by running Nikto with the \-h (\-help) option\&.
.PP
\fB\-Cgidirs\fR
.RS 4
Scan these CGI directories\&. Special words "none" or "all" may be used to scan all CGI directories or none, (respectively)\&. A literal value for a CGI directory such as "/cgi\-test/" may be specified (must include trailing slash)\&. If this is option is not specified, all CGI directories listed in config\&.txt will be tested\&.
.RE
.PP
\fB\-config\fR
.RS 4
Specify an alternative config file to use instead of the config\&.txt located in the install directory\&.
.RE
.PP
\fB\-dbcheck\fR
.RS 4
Check the scan databases for syntax errors\&.
.RE
.PP
\fB\-Display\fR
.RS 4
Control the output that Nikto shows\&. See Chapter 5 for detailed information on these options\&. Use the reference number or letter to specify the type, multiple may be used:
.sp
1 \- Show redirects
.sp
2 \- Show cookies received
.sp
3 \- Show all 200/OK responses
.sp
4 \- Show URLs which require authentication
.sp
D \- Debug Output
.sp
V \- Verbose Output
.RE
.PP
\fB\-evasion\fR
.RS 4
Specify the LibWhisker IDS evasion technique to use (see the LibWhisker docs for detailed information on these)\&. Use the reference number to specify the type, multiple may be used:
.sp
1 \- Random URI encoding (non\-UTF8)
.sp
2 \- Directory self\-reference (/\&./)
.sp
3 \- Premature URL ending
.sp
4 \- Prepend long random string
.sp
5 \- Fake parameter
.sp
6 \- TAB as request spacer
.sp
7 \- Change the case of the URL
.sp
8 \- Use Windows directory separator (\e)
.RE
.PP
\fB\-Format\fR
.RS 4
Save the output file specified with \-o (\-output) option in this format\&. If not specified, the default will be taken from the file extension specified in the \-output option\&. Valid formats are:
.sp
csv \- a comma\-seperated list
.sp
htm \- an HTML report
.sp
txt \- a text report
.sp
xml \- an XML report
.RE
.PP
\fB\-host\fR
.RS 4
Host(s) to target\&. Can be an IP address, hostname or text file of hosts\&. A single dash (\-) maybe used for stdout\&. Can also parse nmap \-oG style output
.RE
.PP
\fB\-Help\fR
.RS 4
Display extended help information\&.
.RE
.PP
\fB\-id\fR
.RS 4
ID and password to use for host Basic host authentication\&. Format is "id:password"\&.
.RE
.PP
\fB\-list\-plugins\fR
.RS 4
Will list all plugins that Nikto can run against targets and then will exit without performing a scan\&. These can be tuned for a session using the \-plugins option\&.
.sp
The output format is:
.sp
Plugin
\fIname\fR
.sp
\ \&\fIfull name\fR
\-
\fIdescription\fR
.sp
\ \&Written by
\fIauthor\fR, Copyright (C)
\fIcopyright\fR
.RE
.PP
\fB\-mutate\fR
.RS 4
Specify mutation technique\&. A mutation will cause Nikto to combine tests or attempt to guess values\&. These techniques may cause a tremendous amount of tests to be launched against the target\&. Use the reference number to specify the type, multiple may be used:
.sp
1 \- Test all files with all root directories
.sp
2 \- Guess for password file names
.sp
3 \- Enumerate user names via Apache (/~user type requests)
.sp
4 \- Enumerate user names via cgiwrap (/cgi\-bin/cgiwrap/~user type requests)
.sp
5 \- Attempt to brute force sub\-domain names, assume that the host name is the parent domain
.sp
6 \- Attempt to guess directory names from the supplied dictionary file
.RE
.PP
\fB\-mutate\-options\fR
.RS 4
Provide extra information for mutates, e\&.g\&. a dictionary file
.RE
.PP
\fB\-nolookup\fR
.RS 4
Do not perform name lookups on IP addresses\&.
.RE
.PP
\fB\-nossl\fR
.RS 4
Do not use SSL to connect to the server\&.
.RE
.PP
\fB\-no404\fR
.RS 4
Disable 404 (file not found) checking\&. This will reduce the total number of requests made to the webserver and may be preferable when checking a server over a slow link, or an embedded device\&. This will generally lead to more false positives being discovered\&.
.RE
.PP
\fB\-output\fR
.RS 4
Write output to the file specified\&. The format used will be taken from the file extension\&. This can be over\-riden by using the \-Format option (e\&.g\&. to write text files with a different extension\&. Existing files will have new information appended\&.
.RE
.PP
\fB\-plugins\fR
.RS 4
Select which plugins will be run on the specified targets\&. A comma separated list should be provided which lists the names of the plugins\&. The names can be found by using \-list\-plugins\&.
.sp
There are two special entries: ALL, which specifies all plugins shall be run and NONE, which specifies no plugins shall be run\&. The default is ALL
.RE
.PP
\fB\-port\fR
.RS 4
TCP port(s) to target\&. To test more than one port on the same host, specify the list of ports in the \-p (\-port) option\&. Ports can be specified as a range (i\&.e\&., 80\-90), or as a comma\-delimited list, (i\&.e\&., 80,88,90)\&. If not specified, port 80 is used\&.
.RE
.PP
\fB\-Pause\fR
.RS 4
Seconds to delay between each test\&.
.RE
.PP
\fB\-root\fR
.RS 4
Prepend the value specified to the beginning of every request\&. This is useful to test applications or web servers which have all of their files under a certain directory\&.
.RE
.PP
\fB\-ssl\fR
.RS 4
Only test SSL on the ports specified\&. Using this option will dramatically speed up requests to HTTPS ports, since otherwise the HTTP request will have to timeout first\&.
.RE
.PP
\fB\-Single\fR
.RS 4
Perform a single request to a target server\&. Nikto will prompt for all options which can be specified, and then report the detailed output\&. See Chapter 5 for detailed information\&.
.RE
.PP
\fB\-timeout\fR
.RS 4
Seconds to wait before timing out a request\&. Default timeout is 10 seconds\&.
.RE
.PP
\fB\-Tuning\fR
.RS 4
Tuning options will control the test that Nikto will use against a target\&. By default, if any options are specified, only those tests will be performed\&. If the "x" option is used, it will reverse the logic and exclude only those tests\&. Use the reference number or letter to specify the type, multiple may be used:
.sp
0 \- File Upload
.sp
1 \- Interesting File / Seen in logs
.sp
2 \- Misconfiguration / Default File
.sp
3 \- Information Disclosure
.sp
4 \- Injection (XSS/Script/HTML)
.sp
5 \- Remote File Retrieval \- Inside Web Root
.sp
6 \- Denial of Service
.sp
7 \- Remote File Retrieval \- Server Wide
.sp
8 \- Command Execution / Remote Shell
.sp
9 \- SQL Injection
.sp
a \- Authentication Bypass
.sp
b \- Software Identification
.sp
c \- Remote Source Inclusion
.sp
x \- Reverse Tuning Options (i\&.e\&., include all except specified)
.sp
The given string will be parsed from left to right, any x characters will apply to all characters to the right of the character\&.
.RE
.PP
\fB\-useproxy\fR
.RS 4
Use the HTTP proxy defined in the configuration file\&.
.RE
.PP
\fB\-update\fR
.RS 4
Update the plugins and databases directly from cirt\&.net\&.
.RE
.PP
\fB\-Version\fR
.RS 4
Display the Nikto software, plugin and database versions\&.
.RE
.PP
\fB\-vhost\fR
.RS 4
Specify the Host header to be sent to the target\&.
.RE
.SH "FILES"
.PP
\fInikto\&.conf\fR
.RS 4
The Nikto configuration file\&. This sets Nikto\'s global options\&. Several nikto\&.conf files may exist and are parsed in the below order\&. As each configuration file is loaded is supersedes any previously set configuration:
.sp
.RS 4
\h'-04'\(bu\h'+03'System wide (e\&.g\&. /etc/nikto\&.conf)
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Home directory (e\&.g\&. $HOME/nikto\&.conf)
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Current directory (e\&.g\&. \&./nikto\&.conf)
.RE
.RE
.PP
\fI${NIKTO_DIR}/plugins/db*\fR
.RS 4
db files are the databases that nikto uses to check for vulnerabilities and issues within the web server\&.
.RE
.PP
\fI${NIKTO_DIR}/plugins/*\&.plugin\fR
.RS 4
All nikto\'s plugins exist here\&. Nikto itself is just a wrapper script to manage CLI and pass through to the plugins\&.
.RE
.PP
\fI${NIKTO_DIR}/templates\fR
.RS 4
Contains the templates for nikto\'s output formats\&.
.RE
.SH "BUGS"
.PP
The current features are not supported:
.sp
.RS 4
\h'-04'\(bu\h'+03'SOCKS Proxies
.RE
.SH "AUTHORS"
.PP
Nikto is written and maintained by Chris Sullo and David Lodge\&. See the main documentation for other contributors\&.
.PP
All code is (C) Chris Sullo\&., except LibWhisker which is (C) rfp\&.labs (wiretrip\&.net)\&. Other portions of code may be (C) as specified\&.
.SH "SEE ALSO"
.PP

\fINikto Homepage\fR\&[1]
.SH "NOTES"
.IP " 1." 4
Nikto Homepage
.RS 4
\%https://cirt.net/
.RE


program/nikto.pl
#!/usr/bin/env perl
use strict;
#VERSION,2.1.6
###############################################################################
# Modules are now loaded in a function so errors can be trapped and evaluated
load_modules();
###############################################################################
#                               Nikto                                         #
###############################################################################
#  Copyright (C) 2001 Chris Sullo
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; version 2
#  of the License only.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to
#  Free Software Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# Contact Information:
#     Sullo (sullo@cirt.net)
#     http://cirt.net/
#######################################################################
# See the COPYING file for more information on the License Nikto is distributed under.
#
# This program is intended for use in an authorized manner only, and the author
# can not be held liable for anything done with this program, code, or items discovered
# with this program's use.
#######################################################################

# global var/definitions
use vars qw/$TEMPLATES %CLI %VARIABLES %TESTS/;
use vars qw/%NIKTO %CONFIGFILE %COUNTERS %db_extensions/;
use vars qw/@RESULTS @PLUGINS @DBFILE @REPORTS %CONTENTSEARCH/;

# setup
Getopt::Long::Configure('no_ignore_case');
$COUNTERS{'scan_start'}  = time();
$VARIABLES{'DIV'}        = "-" x 75;
$VARIABLES{'name'}       = "Nikto";
$VARIABLES{'version'}    = "2.1.6";

# signal trap so we can close down reports properly
$SIG{'INT'} = \&safe_quit;

config_init();
setup_dirs();
require "$CONFIGFILE{'PLUGINDIR'}/nikto_core.plugin";
nprint("T:" . localtime($COUNTERS{'scan_start'}) . ": Starting", "d");
require "$CONFIGFILE{'PLUGINDIR'}/LW2.pm";
$VARIABLES{'GMTOFFSET'} = gmt_offset();

# use LW2;                   ### Change this line to use a different installed version

#set SSL Engine
LW2::init_ssl_engine($CONFIGFILE{'LW_SSL_ENGINE'});

my ($a, $b) = split(/\./, $LW2::VERSION);
die("- You must use LW2 2.4 or later\n") if ($a != 2 || $b < 4);

general_config();
load_databases();
load_databases('u');
nprint("- $VARIABLES{'name'} v$VARIABLES{'version'}");
nprint($VARIABLES{'DIV'});

# No targets - quit while we're ahead
if ($CLI{'host'} eq '') {
    nprint("+ ERROR: No host (-host) specified");
    usage(1);
}

$COUNTERS{'total_targets'} = $COUNTERS{'hosts_completed'} = 0;
load_plugins();

my $is_failure = 0;

# Parse the supplied list of targets
my @MARKS = set_targets($CLI{'host'}, $CLI{'ports'}, $CLI{'ssl'}, $CLI{'root'});

if (defined($CLI{'key'}) || defined($CLI{'cert'})) {
    $CLI{'key'}  = $CLI{'cert'} unless (defined($CLI{'key'}));
    $CLI{'cert'} = $CLI{'key'}  unless (defined($CLI{'cert'}));
}

# Open reporting
report_head($CLI{'format'}, $CLI{'file'});

# Now check each target is real and remove duplicates/fill in extra information
foreach my $mark (@MARKS) {
    $mark->{'messages'} =  ();
    $mark->{'test'} = 1;
    $mark->{'failures'} = 0;

    # Try to resolve the host
    my $msgs;
    ($mark->{'hostname'}, $mark->{'ip'}, $mark->{'display_name'}, $msgs) = resolve($mark->{'ident'});
    if ($msgs ne "") {
	push(@{ $mark->{'messages'} }, $msgs);
	#push ($mark->{'messages'}, $msgs);
    }

    # Skip if we can't resolve the host - we'll error later
    if (!defined $mark->{'ip'}) {
        $mark->{'test'} = 0;
        next;
    }

    # Read cookies from conf & set into the cookie jar
    if (defined $CONFIGFILE{'STATIC-COOKIE'}) {
        $mark->{'cookiejar'} = LW2::cookie_new_jar();
        foreach my $p (split(/;/, $CONFIGFILE{'STATIC-COOKIE'})) {
            if ($p =~ /"([^=]+)=(.+)"/) {
                LW2::cookie_set(\%{ $mark->{'cookiejar'} }, $1, $2);
                }
       }
    }

    if (defined $CLI{'vhost'}) { $mark->{'vhost'} = $CLI{'vhost'} }

    # Check that the port is open
    my $open =
      port_check(time(), $mark->{'hostname'}, $mark->{'ip'}, $mark->{'port'}, $CLI{'key'}, $CLI{'cert'}, $mark->{'vhost'});
    if ($open == 0) {
        $mark->{'test'} = 0;
        next;
    }
    else {
        $COUNTERS{'total_targets'}++;
    }
    $mark->{'ssl'} = $open - 1;

    if ($mark->{'ssl'}) {
        $mark->{'key'}  = $CLI{'key'};
        $mark->{'cert'} = $CLI{'cert'};
    }
}

# Load db_tests
set_scan_items();

# Start hook to allow plugins to load databases etc
run_hooks("", "start");

# Now we've done the precursor, do the scan
foreach my $mark (@MARKS) {
    next unless ($mark->{'test'});
    $mark->{'start_time'} = time();
    $VARIABLES{'TEMPL_HCTR'}++;

    if (defined $CLI{'vhost'}) {
        $mark->{'vhost'} = $CLI{'vhost'};
    }

    # Saving responses
    if ($CLI{'saveresults'} ne '') {
        $mark->{'save_dir'} = save_createdir($CLI{'saveresults'}, $mark);
        $mark->{'save_prefix'} = save_getprefix($mark);
    }

    $mark->{'total_vulns'}  = 0;
    $mark->{'total_errors'} = 0;

    my %FoF = ();

    nfetch($mark, "/", "GET", "", "", { noprefetch => 1, nopostfetch => 1 }, "getinfo");

    report_host_start($mark);
    if ($CLI{'findonly'}) {
        my $protocol = "http";
        if ($mark->{'ssl'}) { $protocol .= "s"; }
        if ($mark->{'banner'} eq "") {
            $mark->{'banner'} = "(no identification possible)";
        }

        add_vulnerability($mark,
                   "Server: $protocol://$mark->{'display_name'}:$mark->{'port'}\t$mark->{'banner'}",
                   0);
    }
    else {
        dump_target_info($mark);
        unless ((defined $CLI{'nofof'}) || ($CLI{'plugins'} eq '@@NONE')) { map_codes($mark) }
        run_hooks($mark, "recon");
        run_hooks($mark, "scan");
    }

    if ($mark->{'total_errors'} > 0 || $mark->{'total_vulns'} > 0) {
        $is_failure = 1;
    }

    $mark->{'end_time'} = time();
    $mark->{'elapsed'}  = $mark->{'end_time'} - $mark->{'start_time'};
    if (!$CLI{'findonly'}) {
        if (!$mark->{'terminate'}) {
            nprint(
                "+ $COUNTERS{'totalrequests'} requests: $mark->{'total_errors'} error(s) and $mark->{'total_vulns'} item(s) reported on remote host"
                );
        }
        else {
            nprint(
                "+ SCAN TERMINATED:  $mark->{'total_errors'} error(s) and $mark->{'total_vulns'} item(s) reported on remote host"
                );
                $is_failure = 1;
        }
        nprint(  "+ End Time:           "
               . date_disp($mark->{'end_time'})
               . " (GMT$VARIABLES{'GMTOFFSET'}) ($mark->{'elapsed'} seconds)");
    }
    nprint($VARIABLES{'DIV'});

    $COUNTERS{'hosts_completed'}++;
    report_host_end($mark);
}
$COUNTERS{'scan_end'}     = time();
$COUNTERS{'scan_elapsed'} = ($COUNTERS{'scan_end'} - $COUNTERS{'scan_start'});
report_summary();
report_close();

if (!$CLI{'findonly'}) {
    nprint("+ $COUNTERS{'hosts_completed'} host(s) tested");
    nprint("+ $COUNTERS{'totalrequests'} requests made in $COUNTERS{'scan_elapsed'} seconds", "v");

    send_updates(@MARKS);
}

nprint("T:" . localtime() . ": Ending", "d");

exit $is_failure;

#################################################################################
# Load config files in order
sub config_init {
    my (@CF, $home);
    my $config_exists = 0;

    # read just the --config option
    {
        my %optcfg;
        Getopt::Long::Configure('pass_through', 'noauto_abbrev');
        GetOptions(\%optcfg, "config=s");
        Getopt::Long::Configure('nopass_through', 'auto_abbrev');
        if (defined $optcfg{'config'}) { $VARIABLES{'configfile'} = $optcfg{'config'}; }
    }

    # Guess Nikto current directory
    my $NIKTODIR = abs_path($0);
    chomp($NIKTODIR);
    $NIKTODIR =~ s#[\\/]nikto.pl$##;

    # Guess user's home directory -- to support Windows
    foreach my $var (split(/ /, "HOME USERPROFILE")) {
        $home = $ENV{$var} if ($ENV{$var});
    }

    # Read the conf files in order (previous values are over-written with each, if multiple found)
    push(@CF,"$NIKTODIR/nikto.conf.default");
    push(@CF,"/etc/nikto.conf");
    push(@CF,"$home/nikto.conf");
    push(@CF,"$NIKTODIR/nikto.conf");
    push(@CF,"nikto.conf");
    push(@CF,"$VARIABLES{'configfile'}");

    # load in order, over-writing values as we go
    for (my $i=0;$i<=$#CF;$i++) {
        my $error = load_config($CF[$i]);
        $config_exists = 1 if ($error eq ""); # any loaded is good
    }

    # Couldn't find any
    if ($config_exists == 0) {
        die "- Could not find a valid nikto config file. Tried: @CF\n";
    }

    return;
}

###############################################################################
sub load_modules {
        my $errors=0;
	my @modules = qw/Getopt::Long Time::Local IO::Socket Net::hostent/;
	push(@modules,"List::Util qw(sum)");
	push(@modules,"Cwd 'abs_path'");
	foreach my $mod (@modules) {
		eval "use $mod";
        	if ($@) {
			print "ERROR: Required module not found: $mod\n";
			$errors=1;
		}
	}

	@modules = ();
	push(@modules,"Time::HiRes qw(sleep ualarm gettimeofday tv_interval)");
	push(@modules,"POSIX qw(:termios_h)");
	foreach my $mod (@modules) {
		eval "use $mod";
		if ($@ && $^O !~ /MSWin32/) {
			# Allow this to work on Windows
			if ($@) { print "ERROR: Required module not found: $mod\n"; $errors=1; }
		}
	}

	if ($errors) { exit 1; }
}

#################################################################################
# load config file
# error=load_config(FILENAME)
sub load_config {
    my $configfile = $_[0];

    open(CONF, "<$configfile") || return "+ ERROR: Unable to open config file '$configfile'";
    my @CONFILE = <CONF>;
    close(CONF);

    foreach my $line (@CONFILE) {
        $line =~ s/\#.*$//;
        chomp($line);
        $line =~ s/\s+$//;
        $line =~ s/^\s+//;
        next if ($line eq "");
        my @temp = split(/=/, $line, 2);
        if ($temp[0] ne "") { $CONFIGFILE{ $temp[0] } = $temp[1]; }
    }

    # add CONFIG{'CLIOPTS'} to ARGV if defined...
    if (defined $CONFIGFILE{'CLIOPTS'}) {
        my @t = split(/ /, $CONFIGFILE{'CLIOPTS'});
        foreach my $c (@t) { push(@ARGV, $c); }
    }

    # Check for necessary config items
    check_config_defined("CHECKMETHODS", "HEAD");
    check_config_defined('@@DEFAULT',    '@@ALL');

    return "";
}
#################################################################################
# find plugins directory
sub setup_dirs {
    my $CURRENTDIR = abs_path($0);
    chomp($CURRENTDIR);
    $CURRENTDIR =~ s#[\\/]nikto.pl$##;
    $CURRENTDIR = "." if $CURRENTDIR =~ /^nikto.pl$/;

    # First assume we get it from CONFIGFILE
    unless (defined $CONFIGFILE{'EXECDIR'}) {
        if (-d "$ENV{'PWD'}/plugins") {
            $CONFIGFILE{'EXECDIR'} = $ENV{'PWD'};
        }
        elsif (-d "$CURRENTDIR/plugins") {
            $CONFIGFILE{'EXECDIR'} = $CURRENTDIR;
        }
        elsif (-d "./plugins") {
            $CONFIGFILE{'EXECDIR'} = $CURRENTDIR;
        }
        else {
            print STDERR "Could not work out the nikto EXECDIR, try setting it in nikto.conf\n";
            exit;
        }
    }
    unless (defined $CONFIGFILE{'PLUGINDIR'}) {
        $CONFIGFILE{'PLUGINDIR'} = "$CONFIGFILE{'EXECDIR'}/plugins";
    }
    unless (defined $CONFIGFILE{'TEMPLATEDIR'}) {
        $CONFIGFILE{'TEMPLATEDIR'} = "$CONFIGFILE{'EXECDIR'}/templates";
    }
    unless (defined $CONFIGFILE{'DOCDIR'}) {
        $CONFIGFILE{'DOCDIR'} = "$CONFIGFILE{'EXECDIR'}/docs";
    }
    unless (defined $CONFIGFILE{'DBDIR'}) {
        $CONFIGFILE{'DBDIR'} = "$CONFIGFILE{'EXECDIR'}/databases";
    }
    return;
}

######################################################################
## check_config_defined(item, default)
## Checks whether config has been set, warns and sets to a default
sub check_config_defined {
    my $item    = $_[0];
    my $default = $_[1];

    if (!defined $CONFIGFILE{$item}) {
        print STDERR
          "- Warning: $item is not defined in Nikto configuration, setting to \"$default\"\n";
        $CONFIGFILE{$item} = $default;
    }

    return;
}
